### PR TITLE
[RUM-5918] Update native Datadog SDKs

### PR DESCRIPTION
**Pull request by marco-saia-datadog**
### What does this PR do?

Bumps native Datadog SDKs versions:

* **iOS** 2.14.1 -> 2.18.0
* **Android** 2.11.0 -> 2.14.0

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)
- [ ] If this PR is auto-generated, please make sure also to manually update the code related to the change

### CODE REVIEW

File: packages/core/android/gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.5-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.0-bin.zip
```
Comment (0xnm) [inline packages/core/android/gradle/wrapper/gradle-wrapper.properties]: when updating Gradle it is better to use `./gradlew wrapper --gradle-version 8.0 --distribution-type all` command. Advantages:

* It will also update the wrapper script as well
* By supplying `all` distribution type you will keep the distribution with all the javadoc included, which is handy.

Same comment applies to other Gradle updates in this PR.

File: example-new-architecture/android/build.gradle [hunk 2]
```diff
@@ -7,7 +7,7 @@ buildscript {
         compileSdkVersion = 34
         targetSdkVersion = 34
         ndkVersion = "25.1.8937393"
-        kotlinVersion = "1.8.0"
+        kotlinVersion = "1.9.0"
```
Comment (0xnm) [inline example-new-architecture/android/build.gradle]: so why doesn't it compile with Kotlin 1.8 (I know that Android SDK moved to Kotlin 1.9) - which error exactly are you getting?

Comment (marco-saia-datadog) [reply example-new-architecture/android/build.gradle]: I think I can actually compile with Kotlin 1.8.21 and Gradle 7.3.3. I am running some tests on the CI with the current setup, I'll then re-run it with the new values and see if it works.

I was getting conflicting kotlin versions (duplicated kotlin-stdlib related dependencies conflict)

File: packages/core/android/gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (marco-saia-datadog) [reply packages/core/android/gradle/wrapper/gradle-wrapper.properties]: Sounds good! Thank you

File: example-new-architecture/android/build.gradle (hunk 2 above)
Comment (0xnm) [reply example-new-architecture/android/build.gradle]: Like this one? https://github.com/DataDog/dd-sdk-android/blob/fd7786e2b8cc5513fc441b76dd6dcf1c98de6d9c/MIGRATION.MD?plain=1#L36

It is better to use lower Kotlin version if possible, indeed.

**Comment by fuzzybinary** [comment]
You might want to wait on this for 2.13.1.

File: example-new-architecture/android/build.gradle [hunk 3]
```diff
@@ -3,11 +3,11 @@
 buildscript {
     ext {
         buildToolsVersion = "34.0.0"
-        minSdkVersion = 21
+        minSdkVersion = 23
```
Comment (0xnm) [inline example-new-architecture/android/build.gradle:6]: why do we raise min SDK? Is it because of the RN version used in this sample app?

File: packages/core/android/build.gradle [hunk 4]
```diff
@@ -107,14 +107,13 @@ android {
...
-  if (agpVersion.tokenize('.')[0].toInteger() < 8) {
-    compileOptions {
-      sourceCompatibility JavaVersion.VERSION_11
-      targetCompatibility JavaVersion.VERSION_11
-    }
-    kotlinOptions {
-      jvmTarget = JavaVersion.VERSION_11.majorVersion
-    }
+
+  compileOptions {
```
Comment (0xnm) [inline packages/core/android/build.gradle]: what is the reason for this change?

File: packages/react-native-session-replay/android/gradle.properties [hunk 5]
```diff
@@ -1,4 +1,4 @@
-DatadogSDKReactNativeSessionReplay_kotlinVersion=1.7.21
+DatadogSDKReactNativeSessionReplay_kotlinVersion=1.9.0
```
Comment (0xnm) [inline packages/react-native-session-replay/android/gradle.properties]: should it be `1.8.21` here?

File: example-new-architecture/android/build.gradle (hunk 3 above)
Comment (marco-saia-datadog) [reply example-new-architecture/android/build.gradle:6]: I finally got back to this again. Sorry for the long wait. Yes, a bump of the version was required for new architecture in the example app.

File: packages/core/android/build.gradle (hunk 4 above)
Comment (marco-saia-datadog) [reply packages/core/android/build.gradle]: All versions of RN that we support can be compiled with Java 17, and I had compatibility problems after bumping the Kotlin version and updating Gradle tools versions. I think we can safely drop support for Java 11 at this point, but I am happy to re-check if you believe this could be an issue.

**Comment by marco-saia-datadog** [comment]
New versions:

- **iOS** 2.14.1 -> 2.18.0
- **Android** 2.11.0 -> 2.14.0

File: packages/core/android/build.gradle (hunk 4 above)
Comment (0xnm) [reply packages/core/android/build.gradle]: so what is the error you are getting?

File: example/android/build.gradle [hunk 6]
```diff
@@ -1,11 +1,11 @@
 buildscript {
     ext {
         buildToolsVersion = "34.0.0"
-        minSdkVersion = 21
+        minSdkVersion = 23
```
Comment (0xnm) [inline example/android/build.gradle:4]: is it because of the latest RN version?

File: packages/core/android/src/main/kotlin/com/datadog/reactnative/DdSdkSynthetics.kt [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.reactnative
+
+internal object DdSdkSynthetics {
+    var testId: String? = null
+    var resultId: String? = null
+}
```
Comment (0xnm) [inline packages/core/android/src/main/kotlin/com/datadog/reactnative/DdSdkSynthetics.kt:12]:
```suggestion
}

```

File: packages/core/android/gradle.properties [hunk 8]
```diff
@@ -1,4 +1,4 @@
-DdSdkReactNative_kotlinVersion=1.7.21
+DdSdkReactNative_kotlinVersion=1.8.21
```
Comment (0xnm) [inline packages/core/android/gradle.properties:1]: what is the error you are getting if Kotlin 1.7 is kept?

**Review by 0xnm — APPROVED** [review]

File: example/android/build.gradle (hunk 6 above)
Comment (marco-saia-datadog) [reply example/android/build.gradle:4]: Yes `com.facebook.react:react-android:0.74.3` requires a minimum `minSdkVersion` of 23.

File: packages/core/android/gradle.properties (hunk 8 above)
Comment (marco-saia-datadog) [reply packages/core/android/gradle.properties:1]: After all the changes, reverting back to 1.7.21 does not seem to cause any issues now. I will look for the cause, and in the meantime I am running tests on the CI for all versions to make sure everything still works fine.

Comment (marco-saia-datadog) [reply packages/core/android/gradle.properties:1]: For reference, 1.7.21 breaks our internal core package build and I could not identify the cause just yet. Since it is not a blocker, I will consider this done for now, by keeping v1.8.21, and keep a note for investigating this further in the future.

**Review by 0xnm — APPROVED** [review]

**Comment by compwron** [comment]
My team is excited for this :)
